chapters/en/chapter10/4.mdx

```diff
@@ -0,0 +1,343 @@
+
+print(f"Mean: {np.mean(sample['array']):.3}, Variance: {np.var(sample['array']):.3}")
+```
+
+```
+Mean: -2.17e-05, Variance: 0.00379
+```
+We can see that the mean is close to zero already, but the standard deviation is a long way off! This would cause our model problems, as the dynamic range of the audio data would be very small and difficult to separate. Let's apply the feature extractor and see what the outputs look like:
+
+```python
+inputs = feature_extractor(sample["array"], sampling_rate=sample["sampling_rate"])
+
+print(f"inputs keys: {list(inputs.keys())}")
+
+print(f"Mean: {np.mean(inputs['input_values']):.3}, Variance: {np.var(inputs['input_values']):.3}")
+```
+
+```
+inputs keys: ['input_values', 'attention_mask']
+Mean: -2.8e-09, Variance: 1.0
+```
+Alright! Our feature extractor returns a dictionary of two arrays: `input_values` and `attention_mask`. The `input_values` are the preprocessed audio inputs that we'd pass to the Wav2Vec2 model. The `attention_mask` would be used if we processed a _batch_ of audio inputs at once.
+
+We can see that the mean value is now very much closer to zero, and the variance bang-on one! This is exactly the form we want our audio samples in prior to feeding them to the Wav2Vec2 encoder.
+
+Note how we've passed the sampling rate of our audio data to our feature extractor. This is good practice, as the feature extractor performs a check under-the-hood to make sure the sampling rate of our audio data matches the sampling rate expected by the model.
+
+
+Let's combine our feature extractor and tokenizer into one function to jointly preprocess our audio and text data:
+
+```python
+def prepare_dataset(batch):
+    # process audio
+    sample = batch["audio"]
+    inputs = feature_extractor(sample["array"], sampling_rate=sample["sampling_rate"])
+    # process audio length
+    batch["input_values"] = inputs.input_values[0]
+    batch["input_length"] = len(batch["input_values"])
+
+    # process targets
+    input_str = batch["text"].lower()
+    # tokenize
+    batch["labels"] = tokenizer(input_str).input_ids
+    return batch
+
+
+vectorized_datasets = raw_datasets.map(
+    prepare_dataset,
+    remove_columns=next(iter(raw_datasets.values())).column_names,
+    desc="preprocess dataset",
+)
+```
+
+Note that it is important to align the decoder's vocabulary with the speech transcriptions of the dataset.
+
+E.g. Librispeech: only captilised letters in the transcriptions, whereas BART was pretrained mostly on lower-cased text.
+
+Filter to max len 20s
+
+```python
+MAX_INPUT_LENGTH_IN_SECONDS = 20
+
+max_input_length = MAX_INPUT_LENGTH_IN_SECONDS * feature_extractor.sampling_rate
+# filter data that is longer than max_input_length
+def is_audio_in_length_range(length):
+    return length < max_input_length
+
+vectorized_datasets = vectorized_datasets.filter(
+    is_audio_in_length_range,
+    input_columns=["input_length"],
+)
+```
+
+Define the datacollator:
```

Comment:
Do we really need to roll our own collator for this task? Would be nice if this was natively included in `transformers` right?

If that's not possible, please make sure you explain why we need this collator and also remove the docstring to keep the code consise